In [5]:
CENSUS_TABLES_TO_GET = ['B25041', 'B15003', 'B25063', 'B03002', 'B19001', 'B17017', 'B19301', 'B19057', 'B22010', 'B19056', 'B25003', 'B25075']

# from eg https://censusreporter.org/data/table/?table=B25041&geo_ids=15000US250250909011&primary_geo_id=15000US250250909011
CENSUS_TABLE_MAPPING = {
    'B25041': 'Bedrooms',
    'B15003': 'Educational Attainment for the Population 25 Years and Over',
    'B25063': 'Gross Rent',
    'B03002': 'Hispanic or Latino Origin by Race',
    'B19001': 'Household Income in the Past 12 Months (In 2015 Inflation-adjusted Dollars)',
    'B17017': 'Poverty Status in the Past 12 Months by Household Type by Age of Householder',
    'B19301': 'Per Capita Income in the Past 12 Months (In 2015 Inflation-adjusted Dollars)',
    'B19057': 'Public Assistance Income in the Past 12 Months for Households',
    'B22010': 'Receipt of Food Stamps/SNAP in the Past 12 Months by Disability Status for Households',
    'B19056': 'Supplemental Security Income (SSI) in the Past 12 Months for Households',
    'B25003': 'Tenure',
    'B25075': 'Value'
}

CENSUS_TABLE_MY_MAPPING = {
    'B25041': 'Bedrooms',
    'B15003': 'school_',
    'B25063': 'rent_',
    'B03002': 'Hispanic or Latino Origin by Race',
    'B19001': 'income_',
    'B17017': 'poverty_',
    'B19301': 'income_per_capita_',
    'B19057': 'income_public_assistance_',
    'B22010': 'food_stamps_',
    'B19056': 'ssi_',
    'B25003': 'rent_or_buy_',
    'B25075': 'value_'
}

In [1]:
import pandas as pd
from IPython.display import display

In [2]:
def process_columns(columns, table_id, mapping):
    new_cols = []
    
    for col in columns:
        if 'Estimate' in col:
            if col[19:] == '':
                col = '_' * 19 + 'total'
            
            if mapping:
                col = "{table_name} {num_range}".format(
                    table_name=mapping[table_id],
                    num_range=col[19:].replace('$', '')
                )
            else:
                col = col[19:].replace('$', '')
        
        new_cols += [col]
            
    return new_cols

def preprocess_census_csv(table_id, mapping=None):
    file_path = '../../data/census-data/ACS_15_5YR_{}_with_ann.csv'.format(table_id)
    df = pd.read_csv(file_path, header=1)
    df = df[[col for col in df.columns if 'Margin' not in col]]
    df.insert(0, 'tract_and_block_group', df['Id2'].apply(lambda id_: str(id_)[-7:]))
    df = df.drop(['Id', 'Id2', 'Geography'], axis=1)
    df.columns = process_columns(df.columns, table_id, mapping)
    return df

## Manual transformations for ea table bc I'm too lazy to write up mapping dictionary

In [11]:
df0 = preprocess_census_csv(CENSUS_TABLES_TO_GET[0])
df0.columns = ['tract_and_block_group', 'bedroom_total_ppl', 'bedroom_0', 'bedroom_1', 'bedroom_2', 'bedroom_3', 'bedroom_4', 'bedroom_5+']
df0.head(1).T

0
tract_and_block_group  0001001
bedroom_total_ppl          560
bedroom_0                    8
bedroom_1                  120
bedroom_2                  193
bedroom_3                  183
bedroom_4                   56
bedroom_5+                   0

In [12]:
df1 = preprocess_census_csv(CENSUS_TABLES_TO_GET[1])
df1.columns = ['tract_and_block_group',
 'school_total',
 'school_0_none',
 'school_1_preschool',
 'school_2_kindergarden',
 'school_3_1st_grade',
 'school_4_2nd_grade',
 'school_5_3rd_grade',
 'school_6_4th_grade',
 'school_7_5th_grade',
 'school_8_6th_grade',
 'school_9_7th_grade',
 'school_10_8th_grade',
 'school_11_9th_grade',
 'school_12_10th_grade',
 'school_13_11th_grade',
 'school_14_12th_grade_no_diploma',
 'school_15_hs_diploma',
 'school_16_ged',
 'school_17_less_than_1_yr',
 'school_18_some_college_no_degree',
 "school_19_associates",
 "school_20_bachelors",
 "school_21_masters",
 'school_22_professional_school',
 'school_23_doctorate']
df1.head(1).T

0
tract_and_block_group             0001001
school_total                          960
school_0_none                          23
school_1_preschool                      0
school_2_kindergarden                   0
school_3_1st_grade                      0
school_4_2nd_grade                      0
school_5_3rd_grade                      9
school_6_4th_grade                      0
school_7_5th_grade                      0
school_8_6th_grade                      0
school_9_7th_grade                      0
school_10_8th_grade                   142
school_11_9th_grade                     0
school_12_10th_grade                    1
school_13_11th_grade                   12
school_14_12th_grade_no_diploma        34
school_15_hs_diploma                  281
school_16_ged                          57
school_17_less_than_1_yr               19
school_18_some_college_no_degree       37
school_19_associates                    0
school_20_bachelors                   298
school_21_masters                      36
school_22_professional_school          11
school_23_doctorate                     0

In [13]:
df2.head(1).drop(1, axis=1).T

NameError: name 'df2' is not defined

In [14]:
df2 = preprocess_census_csv(CENSUS_TABLES_TO_GET[2])
df2 = df2.drop(['total', 'ent'], axis=1)
df2.columns = ['tract_and_block_group',
 'rent_total',
 'rent_0_100',
 'rent_100_149',
 'rent_150_199',
 'rent_200_249',
 'rent_250_299',
 'rent_300_349',
 'rent_350_399',
 'rent_400_449',
 'rent_450_499',
 'rent_500_549',
 'rent_550_599',
 'rent_600_649',
 'rent_650_699',
 'rent_700_749',
 'rent_750_799',
 'rent_800_899',
 'rent_900_999',
 'rent_1000_1249',
 'rent_1250_1499',
 'rent_1500_1999',
 'rent_2000_2499',
 'rent_2500_2999',
 'rent_3000_3499',
 'rent_3500+']
df2.head(1).T

0
tract_and_block_group  0001001
rent_total                 348
rent_0_100                   0
rent_100_149                 0
rent_150_199                 0
rent_200_249                36
rent_250_299                40
rent_300_349                 9
rent_350_399                 9
rent_400_449                12
rent_450_499                13
rent_500_549                 0
rent_550_599                 0
rent_600_649                 8
rent_650_699                51
rent_700_749                 0
rent_750_799                 0
rent_800_899                 0
rent_900_999                 0
rent_1000_1249              41
rent_1250_1499               9
rent_1500_1999              37
rent_2000_2499              73
rent_2500_2999               0
rent_3000_3499               0
rent_3500+                  10

In [3]:
def process_columns(columns, table_id, mapping):
    return columns

In [6]:
df3 = preprocess_census_csv(CENSUS_TABLES_TO_GET[3])
df3 = df3.iloc[:, :13]
df3 = df3[[i for i in df3.columns if not ('Two' in i or 'other' in i.lower() or 'Indian' in i)]]
df3.columns = ['tract_and_block_group',
 'race_total',
 'race_non_hispanic_total',
 'race_white',
 'race_black',
 'race_asian',
 'race_hispanic']
df3['race_other'] = df3['race_non_hispanic_total'] - df3['race_white'] - df3['race_black'] - df3['race_asian']
df3 = df3.drop('race_non_hispanic_total', axis=1)
df3.head(1).T

0
tract_and_block_group  0001001
race_total                1374
race_white                 593
race_black                 126
race_asian                 322
race_hispanic              316
race_other                  17

In [8]:
df3[df3.tract_and_block_group.str.contains('981700')]

tract_and_block_group  race_total  race_white  race_black  race_asian  \
643               9817001           0           0           0           0   

     race_hispanic  race_other  
643              0           0

In [17]:
def process_columns(columns, table_id, mapping):
    new_cols = []
    
    for col in columns:
        if 'Estimate' in col:
            if col[19:] == '':
                col = '_' * 19 + 'total'
            
            if mapping:
                col = "{table_name}{num_range}".format(
                    table_name=mapping[table_id],
                    num_range=col[19:].
                        replace('$', '').
                        replace(' to ', '_').
                        replace(',', '').
                        replace('Less than ', '0_').
                        replace(' or more', '+')
                )
            else:
                col = col[19:].replace('$', '')
        
        new_cols += [col]
            
    return new_cols

In [18]:
df4 = preprocess_census_csv(CENSUS_TABLES_TO_GET[4], CENSUS_TABLE_MY_MAPPING)
df4.head(1).T

0
tract_and_block_group  0001001
income_total               543
income_0_10000              24
income_10000_14999         132
income_15000_19999          16
income_20000_24999          30
income_25000_29999           3
income_30000_34999          83
income_35000_39999           0
income_40000_44999          16
income_45000_49999          13
income_50000_59999          72
income_60000_74999          23
income_75000_99999          44
income_100000_124999        16
income_125000_149999        15
income_150000_199999         0
income_200000+              56

In [19]:
df5 = preprocess_census_csv(CENSUS_TABLES_TO_GET[5], CENSUS_TABLE_MY_MAPPING)
df5 = df5.iloc[:, :3]
df5.columns = ['tract_and_block_group',
 'poverty_total_pop',
 'poverty_pop_below_poverty_level']
df5.head(1).T

0
tract_and_block_group            0001001
poverty_total_pop                    543
poverty_pop_below_poverty_level      122

In [20]:
df6 = preprocess_census_csv(CENSUS_TABLES_TO_GET[6], CENSUS_TABLE_MY_MAPPING)
df6.columns = ['tract_and_block_group', 'income_per_capita']
df6.head(1).T

0
tract_and_block_group  0001001
income_per_capita        26233

In [21]:
df7 = preprocess_census_csv(CENSUS_TABLES_TO_GET[7], CENSUS_TABLE_MY_MAPPING)
df7 = df7.iloc[:, :3]
df7 = df7.drop('income_public_assistance_total', axis=1)
df7.columns = ['tract_and_block_group',
 'poverty_pop_w_public_assistance']
df7.head(1).T

0
tract_and_block_group            0001001
poverty_pop_w_public_assistance        9

In [22]:
df8 = preprocess_census_csv(CENSUS_TABLES_TO_GET[8], CENSUS_TABLE_MY_MAPPING)
df8 = df8[['tract_and_block_group', 'food_stamps_ received Food Stamps/SNAP in the past 12 months:']]
df8.columns = ['tract_and_block_group',
 'poverty_pop_w_food_stamps']
df8.head(1).T

0
tract_and_block_group      0001001
poverty_pop_w_food_stamps       89

In [23]:
df9 = preprocess_census_csv(CENSUS_TABLES_TO_GET[9], CENSUS_TABLE_MY_MAPPING)
df9 = df9[['tract_and_block_group', 'ssi_With Supplemental Security Income (SSI)']]
df9.columns = ['tract_and_block_group',
 'poverty_pop_w_ssi']
df9.head(1).T

0
tract_and_block_group  0001001
poverty_pop_w_ssi           12

In [24]:
df10 = preprocess_census_csv(CENSUS_TABLES_TO_GET[10], CENSUS_TABLE_MY_MAPPING)
df10.columns = ['tract_and_block_group',
 'housing_total',
 'housing_own',
 'housing_rent']
df10.head(1).T

0
tract_and_block_group  0001001
housing_total              543
housing_own                177
housing_rent               366

In [25]:
df11 = preprocess_census_csv(CENSUS_TABLES_TO_GET[11], CENSUS_TABLE_MY_MAPPING)
# df11 = df11.iloc[:, :3]
# df11.columns = ['tract_and_block_group',
#  'poverty_total_pop',
#  'poverty_pop_below_poverty_level']
df11.head(1).T

0
tract_and_block_group  0001001
value_total                177
value_0_10000                0
value_10000_14999            0
value_15000_19999            0
value_20000_24999            0
value_25000_29999            0
value_30000_34999            0
value_35000_39999            0
value_40000_49999            0
value_50000_59999            0
value_60000_69999            0
value_70000_79999            0
value_80000_89999            0
value_90000_99999            0
value_100000_124999          0
value_125000_149999          0
value_150000_174999          7
value_175000_199999         13
value_200000_249999         38
value_250000_299999         22
value_300000_399999          5
value_400000_499999         51
value_500000_749999         41
value_750000_999999          0
value_1000000_1499999        0
value_1500000_1999999        0
value_2000000+               0

In [26]:
list(df5.columns)

['tract_and_block_group',
 'poverty_total_pop',
 'poverty_pop_below_poverty_level']

## Finally finished making dfs; now to check them (manually)

In [27]:
df8.head(1).T

0
tract_and_block_group      0001001
poverty_pop_w_food_stamps       89

## Now to make the giant df and pickle it

In [28]:
dfs = [df0, df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11]

In [29]:
%time
# http://stackoverflow.com/questions/23668427/pandas-joining-multiple-dataframes-on-columns
df_final = reduce(lambda left, right: pd.merge(left, right, on='tract_and_block_group', how='outer'), dfs)

CPU times: user 20 µs, sys: 5 µs, total: 25 µs
Wall time: 16 µs


In [30]:
df_final.shape

(646, 117)

In [31]:
[i.shape[0] for i in dfs]

[646, 646, 646, 646, 646, 646, 645, 646, 646, 646, 646, 646]

### Just to be good, check the `tract_and_block_group` for the 645 one. Check that it's not in Boston

In [32]:
df6.head(1).T

0
tract_and_block_group  0001001
income_per_capita        26233

In [33]:
from_df6 = set(df6.tract_and_block_group.drop_duplicates())
from_df7 = set(df7.tract_and_block_group.drop_duplicates())

In [34]:
print from_df6.difference(from_df7)
print from_df7.difference(from_df6)

set([])
set(['9811003'])


Upon visual inspection [here](https://censusreporter.org/profiles/14000US25025090901-census-tract-90901-suffolk
-ma/), that Census tract doesn't seem to be in Boston.

## OK, final check, then pickle

Is 9811003 in the groups?

In [38]:
'9811003' in df_final.tract_and_block_group.drop_duplicates().tolist()

True

In [41]:
'9811003' in df5.tract_and_block_group.drop_duplicates().tolist()

True

In [42]:
df_final.head(1).T

0
tract_and_block_group             0001001
bedroom_total_ppl                     560
bedroom_0                               8
bedroom_1                             120
bedroom_2                             193
bedroom_3                             183
bedroom_4                              56
bedroom_5+                              0
school_total                          960
school_0_none                          23
school_1_preschool                      0
school_2_kindergarden                   0
school_3_1st_grade                      0
school_4_2nd_grade                      0
school_5_3rd_grade                      9
school_6_4th_grade                      0
school_7_5th_grade                      0
school_8_6th_grade                      0
school_9_7th_grade                      0
school_10_8th_grade                   142
school_11_9th_grade                     0
school_12_10th_grade                    1
school_13_11th_grade                   12
school_14_12th_grade_no_diploma        34
school_15_hs_diploma                  281
school_16_ged                          57
school_17_less_than_1_yr               19
school_18_some_college_no_degree       37
school_19_associates                    0
school_20_bachelors                   298
...                                   ...
housing_total                         543
housing_own                           177
housing_rent                          366
value_total                           177
value_0_10000                           0
value_10000_14999                       0
value_15000_19999                       0
value_20000_24999                       0
value_25000_29999                       0
value_30000_34999                       0
value_35000_39999                       0
value_40000_49999                       0
value_50000_59999                       0
value_60000_69999                       0
value_70000_79999                       0
value_80000_89999                       0
value_90000_99999                       0
value_100000_124999                     0
value_125000_149999                     0
value_150000_174999                     7
value_175000_199999                    13
value_200000_249999                    38
value_250000_299999                    22
value_300000_399999                     5
value_400000_499999                    51
value_500000_749999                    41
value_750000_999999                     0
value_1000000_1499999                   0
value_1500000_1999999                   0
value_2000000+                          0

[117 rows x 1 columns]

In [45]:
ls -lh ../../data

total 3.4G
drwxrwxr-x 2 ubuntu ubuntu 4.0K Jan 31 22:44 census-data/
-rw-rw-r-- 1 ubuntu ubuntu 607K Jan 31 22:56 census_data_aggregated.pkl
-rw-rw-r-- 1 ubuntu ubuntu 283M Jan 30 00:42 data_subset_w_descs_and_census.pkl
-rw-rw-r-- 1 ubuntu ubuntu 1.1G Jan 30 03:31 data_till_june_2016_w_descs_and_census.h5
-rw-rw-r-- 1 ubuntu ubuntu 2.1G Jan 30 00:43 data_w_descs_and_census.h5
drwxrwxr-x 2 ubuntu ubuntu 4.0K Jan 30 00:35 shape-files/


In [44]:
df_final.to_pickle('../../data/census_data_aggregated.pkl')